src/Internal/Payment/State/AuthenticationRequiredState.php

```diff
@@ -0,0 +1,73 @@
+<?php
+/**
+ * Class AuthenticationRequiredState
+ *
+ * @package WooCommerce\Payments
+ */
+
+namespace WCPay\Internal\Payment\State;
+
+use WCPay\Internal\Proxy\LegacyProxy;
+use WCPay\Internal\Service\CheckoutEncryptionService;
+
+/**
+ * The state, which indicates that the payment processing has been completed.
+ */
+class AuthenticationRequiredState extends AbstractPaymentState {
+
+	/**
+	 * Legacy Proxy.
+	 *
+	 * @var LegacyProxy
+	 */
+	private $legacy_proxy;
+
+	/**
+	 * Checkout Encryption service
+	 *
+	 * @var CheckoutEncryptionService
+	 */
+	private $checkout_encryption_service;
+
+
+	/**
+	 * Class constructor, only meant for storing dependencies.
+	 *
+	 * @param StateFactory              $state_factory Factory for payment states.
+	 * @param LegacyProxy               $legacy_proxy Legacy proxy.
+	 * @param CheckoutEncryptionService $checkout_encryption_service Service for encrypting checkout data.
+	 */
+	public function __construct(
+		StateFactory $state_factory,
+		LegacyProxy $legacy_proxy,
+		CheckoutEncryptionService $checkout_encryption_service
+	) {
+		parent::__construct( $state_factory );
+
+		$this->legacy_proxy                = $legacy_proxy;
+		$this->checkout_encryption_service = $checkout_encryption_service;
+	}
+
+	/**
+	 * Get redirect url.
+	 *
+	 * @return string
+	 */
+	public function get_redirect_url() {
+		$context     = $this->get_context();
+		$intent      = $context->get_intent();
+		$next_action = $intent->get_next_action();
+
+		if ( isset( $next_action['type'] ) && 'redirect_to_url' === $next_action['type'] && ! empty( $next_action['redirect_to_url']['url'] ) ) {
+			return $next_action['redirect_to_url']['url'];
+		}
+
+		return sprintf(
+			'#wcpay-confirm-%s:%s:%s:%s',
+			$context->get_amount() > 0 ? 'pi' : 'si',
+			$context->get_order_id(),
+			$this->checkout_encryption_service->encrypt_client_secret( $intent->get_customer_id(), $intent->get_client_secret() ),
+			$this->legacy_proxy->call_function( 'wp_create_nonce', 'wcpay_update_order_status_nonce' )
```

Comment:
There is another thread (https://github.com/Automattic/woocommerce-payments/pull/7471/#pullrequestreview-1674442396 is the link, but GH is removing the hash/deep link) where I mentioned that some of this is too specific to live within the state. What I mean:

1. The `'#wcpay-confirm-%s:%s:%s:%s'` format is tightly tied to our JS implementation. If we wanted to change it, we'd have some very weird `if()`s deep within the payment process. There should be another layer to handle those. I'm not saying it's the gateway class, but this is probably not the right place.
2. Neither the decryption of the client secret, nor the verification of this nonce will be performed within states, but through a controller that will trigger the state machine. As it will be the job of that controller to verify the nonce, its generation should be close to the verification. Presently, that controller is the gateway class, so we should probably keep this there. Once we add 3DS post-verification to the new payment process, we'll create the proper controller, and that code should fit well there.

With that said, this is a lot of Stripe-specific code to put in the gateway. What do you think that creating the new controller(-like) class now, and putting this code there? Then in the gateway we'd be able to use the service, and we'll be able to use it later as well:

```php
public function new_process_payment( WC_Order $order ) {
	// ...
	$state = $service->process_payment( $order->get_id(), $manual_capture );

	if ( $state instanceof AuthenticationRequiredState ) {
		$redirect = $controller->generate_authentication_url( $state->get_context()->get_intent(), $order->get_id() );

		return [
			'result'   => 'success',
			'redirect' => $redirect,
		];
	}
}
```

Note: The amount wouldn't be needed because we can check whether the intent is a setup or payment intent.

Reply:
I agree with your perspective regarding the stripe related code. Relocating this code might be beneficial in the long run.

Moreover, controllers are mainly designed to receive request data, process it by using services for example, and deliver (return) results. Therefore, I suggest moving this code to a service. Since the `CompletedState` and `ErrorState` also need to return responses, it makes sense to me to move them into the same service. This way, we can manage the business logic more effectively and in a centralized way.

The ideal candidate, IMHO, could be `PaymentProcessingService` where we already have original state processing and parsing results at the end of the cycle could be a nice way to handle it here.